subcerts.go

```diff
@@ -0,0 +1,515 @@
+	}
+
+	// Write the length of the public_key field.
+	binary.BigEndian.PutUint16(serialized[4:], uint16(len(serializedPublicKey)))
+
+	// Write the public key.
+	return append(serialized, serializedPublicKey...), nil
+}
+
+// unmarshalCredential decodes a credential and returns it.
+func unmarshalCredential(serialized []byte) (*credential, error) {
+	// Bytes 0-3 are the validity time field; bytes 4-6 are the length of the
+	// serialized SubjectPublicKeyInfo.
+	if len(serialized) < 6 {
+		return nil, errors.New("credential is too short")
+	}
+
+	// Parse the validity time.
+	validTime := time.Duration(binary.BigEndian.Uint32(serialized)) * time.Second
+
+	// Parse the SubjectPublicKeyInfo.
+	pk, scheme, err := unmarshalSubjectPublicKeyInfo(serialized[6:])
+	if err != nil {
+		return nil, err
+	}
+
+	return &credential{validTime, pk, scheme}, nil
+}
+
+// unmarshalSubjectPublicKeyInfo parses a DER encoded SubjectPublicKeyInfo
+// structure into a public key and its corresponding algorithm.
+func unmarshalSubjectPublicKeyInfo(serialized []byte) (crypto.PublicKey, SignatureScheme, error) {
+	publicKey, err := x509.ParsePKIXPublicKey(serialized)
+	if err != nil {
+		return nil, 0, err
+	}
+
+	switch pk := publicKey.(type) {
+	case *ecdsa.PublicKey:
+		curveName := pk.Curve.Params().Name
+		if curveName == "P-256" {
+			return pk, ECDSAWithP256AndSHA256, nil
+		} else if curveName == "P-384" {
+			return pk, ECDSAWithP384AndSHA384, nil
+		} else if curveName == "P-521" {
+			return pk, ECDSAWithP521AndSHA512, nil
+		} else {
+			return nil, 0, fmt.Errorf("curve %s s not supported", curveName)
+		}
+
+	default:
+		return nil, 0, fmt.Errorf("unsupported delgation key type: %T", pk)
+	}
+}
+
+// getCredentialLen returns the number of bytes comprising the serialized
+// credential that starts at the beginning of the input slice. It returns an
+// error if the input is too short to contain a credential.
+func getCredentialLen(serialized []byte) (int, error) {
+	if len(serialized) < 6 {
+		return 0, errors.New("credential is too short")
+	}
+	// First 4 bytes is the validity time.
+	serialized = serialized[4:]
+
+	// The next 2 bytes are the length of the serialized public key.
+	serializedPublicKeyLen := int(binary.BigEndian.Uint16(serialized))
+	serialized = serialized[2:]
+
+	if len(serialized) < serializedPublicKeyLen {
+		return 0, errors.New("public key of credential is too short")
+	}
+
+	return 6 + serializedPublicKeyLen, nil
+}
+
+// DelegatedCredential stores a credential and its delegation.
+type DelegatedCredential struct {
+	// The serialized form of the credential.
+	Raw []byte
+
+	// Time for which the credential is valid. When this data structure is
+	// serialized, this value is converted to a uint32 representing the duration
+	// in seconds.
+	ValidTime time.Duration
+
+	// The credential public key.
+	PublicKey crypto.PublicKey
+
+	// The credential private key.
+	PrivateKey crypto.PrivateKey
```

Comment:
I think we agreed after talking irl that this field would be removed and returned by NewDC. It's awkward having a field that is silently elided by Marshal -> Unmarshal.

Reply:
This was removed by https://github.com/cloudflare/tls-tris/pull/95/commits/54ec0a4b3583d4e9b0c5360917aca820b576bb9f

Reply:
(That commit is gone now; I smashed all of the changes since review began into a new commit https://github.com/cloudflare/tls-tris/pull/95/commits/a34d64ca3f73b18fa9858ec8f7ad7053fd16aadd.)